gnocchi/rest/aggregates/operations.py

```diff
@@ -32,39 +36,207 @@
+
+    elif isinstance(v1, numpy.ndarray):
+        timestamps = t1
+        granularity = g1
+    elif isinstance(v1, numpy.ndarray):
+        timestamps = t2
+        granularity = g2
+    else:
+        timestamps = t1
+        granularity = g1
+
+    values = binary_operators[op](v1, v2)
+    return granularity, timestamps, values, is_aggregated
+
+
+def handle_aggregate(agg, granularity, timestamps, values, is_aggregated,
+                     references):
     values = numpy.array([AGG_MAP[agg](values, axis=1)]).T
     if values.shape[1] != 1:
         raise RuntimeError("Unexpected resulting aggregated array shape: %s" %
                            values)
     return (granularity, timestamps, values, True)
 
 
+def handle_rolling(agg, granularity, timestamps, values, is_aggregated,
+                   references, window):
+    if window > len(values):
+        raise exceptions.UnAggregableTimeseries(
+            references,
+            "Rolling window '%d' is greater than serie length '%d'" %
+            (window, len(values))
+        )
+
+    # TODO(sileht): make a more optimised version that
+    # compute the data across the whole matrix
+    new_values = None
+    timestamps = timestamps[window - 1:]
+    for ts in values.T:
+        # arogozhnikov.github.io/2015/09/30/NumpyTipsAndTricks2.html
+        stride = ts.strides[0]
+        ts = AGG_MAP[agg](as_strided(
+            ts, shape=[len(ts) - window + 1, window],
+            strides=[stride, stride]), axis=1)
+        if new_values is None:
+            new_values = numpy.array([ts])
+        else:
+            new_values = numpy.append(new_values, [ts], axis=0)
+    return granularity, timestamps, new_values.T, is_aggregated
+
+
+def handle_resample(agg, granularity, timestamps, values, is_aggregated,
+                    references, sampling):
+    # TODO(sileht): make a more optimised version that
+    # compute the data across the whole matrix
+    new_values = None
+    result_timestamps = timestamps
+    for ts in values.T:
+        ts = carbonara.AggregatedTimeSerie.from_data(None, agg, timestamps, ts)
+        try:
+            ts = ts.resample(sampling)
+        except carbonara.UnknownAggregationMethod:
+            raise RuntimeError("Unknown aggregation method: %s" % agg)
+        result_timestamps = ts["timestamps"]
+        if new_values is None:
+            new_values = numpy.array([ts["values"]])
+        else:
+            new_values = numpy.append(new_values, [ts["values"]], axis=0)
+    return sampling, result_timestamps, new_values.T, is_aggregated
+
+
 def handle_aggregation_operator(nodes, granularity, timestamps, initial_values,
                                 is_aggregated, references):
     op = aggregation_operators[nodes[0]]
     agg = nodes[1]
     subnodes = nodes[-1]
     args = nodes[2:-1]
     if agg not in AGG_MAP:
-        raise carbonara.UnknownAggregationMethod(agg)
+        raise RuntimeError("Unknown aggregation method: %s" % agg)
```

Comment:
I understand you want to have the server crash and get a 500, but then, if that's the case, do you really need this check at all?

Reply:
It's just a selt-belt, to show a clear message instead of KeyError, but not really useful.
This is catched by voluptuous already.

Reply:
I have removed useless RuntimeError